### PR TITLE
refactor(sbt): Reimplement the parser

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Sbt manager is becoming regex-heavy which makes it hard to maintain long-term. Therefore, I've reimplemented it using parser library.

## Context

- Ref: #16644

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
@cjtim Would you like to continue with your contribution on the basis of this new implementation? I'm sorry for making some parts of your work obsolete, but unfortunately there is too much regular expressions in Sbt manager.

File: lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt [hunk 1]
```diff
@@ -1,7 +1,8 @@
 val ScalaVersion = "2.12.10"
 val versionExample = "0.0.8"
+val fileVersion = "3.2.1"
 
-version := "3.2.1"
+version := fileVersion
```
Comment (viceice) [inline lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt:5]: why this change?

**Comment by cjtim** [comment]
> @cjtim Would you like to continue with your contribution on the basis of this new implementation? I'm sorry for making some parts of your work obsolete, but unfortunately there is too much regular expressions in Sbt manager.

Please go ahead. This is interesting... 😃

File: lib/modules/manager/sbt/__snapshots__/extract.spec.ts.snap [hunk 2]
```diff
@@ -428,40 +443,6 @@ Object {
 }
 `;
 
-exports[`modules/manager/sbt/extract extractPackageFile() extracts deps when scala version is defined in a variable with ThisBuild scope 1`] = `
```
Comment (viceice) [inline lib/modules/manager/sbt/__snapshots__/extract.spec.ts.snap:431]: why those changes?

File: lib/modules/manager/sbt/extract.spec.ts [hunk 3]
```diff
@@ -186,8 +199,12 @@ describe('modules/manager/sbt/extract', () => {
         ThisBuild / scalaVersion := ScalaVersion
         libraryDependencies += "org.example" %% "bar" % "0.0.2"
       `;
-      expect(extractPackageFile(content)).toMatchSnapshot({
+      expect(extractPackageFile(content)).toMatchObject({
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.spec.ts:202]: can we do those things in a separate PR to better verify ?

File: lib/modules/manager/sbt/extract.ts [hunk 4]
```diff
@@ -1,368 +1,291 @@
...
+  let parsedResult: Ctx | undefined;
+
+  try {
+    parsedResult = scala.query(content, query, {
+      vars: {},
+      deps: [],
+      registryUrls: [MAVEN_REPO],
     });
+  } catch (err) /* istanbul ignore next */ {
+    // no-op
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.ts:278]: looks like a bad idea, at least do a debug message

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/sbt/extract.ts [hunk 5]
```diff
@@ -1,368 +1,291 @@
...
+    q.alt<Ctx>(
+      q.sym<Ctx>('classifier').str(depTypeHandler),
+      q.op<Ctx>('%').sym(depTypeHandler),
+      q.op<Ctx>('%').str(depTypeHandler)
+    )
+  )
+  .handler(depHandler);
+
+const sbtPluginMatch = q
+  .sym<Ctx>('addSbtPlugin')
```
Comment (cjtim) [inline lib/modules/manager/sbt/extract.ts:210]: There's `addCompilerPlugin` should be able to match the plugin too.

https://github.com/renovatebot/renovate/pull/17006/files

File: lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt (hunk 1 above)
Comment (zharinov) [reply lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt:5]: `scalaVersion` can be set via variable defined earlier, so we should support this also for `version` (which translates to `packageFileVersion` in Renovate).

Comment (viceice) [reply lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt:5]: ok, so this case is still supported? just wanted to make sure 🤗

Comment (zharinov) [reply lib/modules/manager/sbt/__fixtures__/scala-version-variable.sbt:5]: Sure, I think I'll return this one back and create simpler separate test for the new case

**Review by rarkins — COMMENTED** [review]
Possible to reduce the number of test changes by doing a refactoring of tests first?

**Comment by zharinov** [comment]
I believe I reduced it to the possible minimum. It turned out, the new parsing alters it's behavior for couple cases:
1. When `scalaVersion` is defined via string literal, Renovate treats it as the dependency on scala library, while it lost this property for variable-mediated definition. Now it's fixed.
2. Another change was related to incomplete definition like `"foo" % "bar" % "1.2.3" % <EOF>` which technically is enough for being parsed as the dependency, so we'll be doing this.

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/sbt/extract.spec.ts [hunk 6]
```diff
@@ -66,6 +66,10 @@ describe('modules/manager/sbt/extract', () => {
     it('extracts deps when scala version is defined in a variable', () => {
       expect(extractPackageFile(sbtScalaVersionVariable)).toMatchSnapshot({
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/manager/sbt/extract.spec.ts:67]: We decided to minimize test code changes in this PR

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/sbt/extract.spec.ts:67]: then revert the other snapshot 😉

**Review by viceice — APPROVED** [review]
